File: benchmark/ts/benchmark_sum.ts
/**
 * Benchmark NumJS sum() performance across various array sizes.
 *
 * Measures execution time for array sizes from 10^2 to 10^7 elements.
 * Results are saved as JSON for comparison with NumPy.
 */

import { writeFileSync, mkdirSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';
import { performance } from 'perf_hooks';

// Import from the built package
// The path is relative to the compiled output in dist/benchmark/ts/
// @ts-expect-error - Types are in dist/index.d.ts but TS can't resolve mjs
import { NDArray, loadWasmModule } from '../../numjs.mjs';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Configuration (match Python exactly)
const ARRAY_SIZES = [1e2, 1e3, 1e4, 1e5, 1e6, 1e7];
const WARMUP_ITERATIONS = 3;
const MIN_ITERATIONS = 10;
const MIN_DURATION_MS = 1000; // 1 second

/**
 * Seeded PRNG (Mulberry32) for reproducibility.
 * This ensures consistent random data across runs.
 */
function createSeededRandom(seed: number): () => number {
  return function () {
    let t = (seed += 0x6d2b79f5);
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

/**
 * Box-Muller transform for normal distribution.
 * Generates numbers similar to numpy.random.randn().
 */
function randn(random: () => number): number {
  const u1 = random();
  const u2 = random();
  return Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
}

interface BenchmarkResult {
  size: number;
  iterations: number;
  times_ms: number[];
  mean_ms: number;
  min_ms: number;
  max_ms: number;
  std_ms: number;
  result: number;
}

/**
 * Benchmark sum() for a given array size.
 */
async function benchmarkSum(size: number): Promise<BenchmarkResult> {
  // Create array with reproducible random data
  const random = createSeededRandom(42);
  const data = Array.from({ length: size }, () => randn(random));

  const arr = await NDArray.fromArray(data);

  try {
    // Warmup runs
    for (let i = 0; i < WARMUP_ITERATIONS; i++) {
      arr.sum();
    }

    // Timed runs - adaptive iteration count
    const times: number[] = [];
    const startTotal = performance.now();
    let iterations = 0;
    let result = 0;

    while (
      iterations < MIN_ITERATIONS ||
      performance.now() - startTotal < MIN_DURATION_MS
    ) {
      const start = performance.now();
      result = arr.sum();
      const end = performance.now();
      times.push(end - start);
      iterations++;
    }

    const meanMs = times.reduce((a, b) => a + b, 0) / times.length;
    const variance =
      times.reduce((sum, t) => sum + (t - meanMs) ** 2, 0) / times.length;
    const stdMs = Math.sqrt(variance);

    // Use reduce instead of spread to avoid stack overflow on large arrays
    const minMs = times.reduce((a, b) => Math.min(a, b), Infinity);
    const maxMs = times.reduce((a, b) => Math.max(a, b), -Infinity);

    return {
      size,
      iterations,
      times_ms: times,
      mean_ms: meanMs,
      min_ms: minMs,
      max_ms: maxMs,
      std_ms: stdMs,
      result,
    };
  } finally {
    arr.dispose();
  }
}

async function main() {
  console.log('NumJS sum() Benchmark');
  console.log('Loading WASM module...');
  await loadWasmModule();

  console.log(
    `Array sizes: ${ARRAY_SIZES.map((s) => `10^${Math.log10(s)}`).join(', ')}`
  );
  console.log('-'.repeat(50));

  const results = {
    library: 'numjs',
    version: '0.1.0',
    config: {
      warmup_iterations: WARMUP_ITERATIONS,
      min_iterations: MIN_ITERATIONS,
      min_duration_ms: MIN_DURATION_MS,
    },
    benchmarks: [] as BenchmarkResult[],
  };

  for (const size of ARRAY_SIZES) {
    const sizeStr = size.toLocaleString().padStart(12);
    process.stdout.write(`Benchmarking size ${sizeStr}... `);

    const result = await benchmarkSum(size);
    results.benchmarks.push(result);

    console.log(
      `Mean: ${result.mean_ms.toFixed(4).padStart(10)} ms  (${result.iterations} iterations)`
    );
  }

  // Output to JSON (use project root benchmark/results, not dist)
  const projectRoot = join(__dirname, '..', '..', '..');
  const outputDir = join(projectRoot, 'benchmark', 'results');
  mkdirSync(outputDir, { recursive: true });
  const outputFile = join(outputDir, 'numjs_results.json');

  writeFileSync(outputFile, JSON.stringify(results, null, 2));

  console.log('-'.repeat(50));
  console.log(`Results saved to: ${outputFile}`);
}

main().catch(console.error);

File: benchmark/python/benchmark_sum.py
#!/usr/bin/env python3
"""
Benchmark NumPy sum() performance across various array sizes.

Measures execution time for array sizes from 10^2 to 10^7 elements.
Results are saved as JSON for comparison with NumJS.
"""

import json
import time
import sys
from pathlib import Path

try:
    import numpy as np
except ImportError:
    print("Error: NumPy is not installed. Run: pip install numpy", file=sys.stderr)
    sys.exit(1)

# Configuration
ARRAY_SIZES = [10**2, 10**3, 10**4, 10**5, 10**6, 10**7]
WARMUP_ITERATIONS = 3
MIN_ITERATIONS = 10
MIN_DURATION_SEC = 1.0  # Run for at least 1 second per size


def benchmark_sum(size: int) -> dict:
    """Benchmark sum() for a given array size."""
    # Create array with reproducible random data
    np.random.seed(42)
    arr = np.random.randn(size).astype(np.float64)

    # Warmup runs (primes CPU caches, etc.)
    for _ in range(WARMUP_ITERATIONS):
        _ = arr.sum()

    # Timed runs - adaptive iteration count
    times = []
    start_total = time.perf_counter()
    iterations = 0
    result = 0.0

    while iterations < MIN_ITERATIONS or (time.perf_counter() - start_total) < MIN_DURATION_SEC:
        start = time.perf_counter()
        result = arr.sum()
        end = time.perf_counter()
        times.append(end - start)
        iterations += 1

    return {
        "size": size,
        "iterations": iterations,
        "times_ms": [t * 1000 for t in times],
        "mean_ms": (sum(times) / len(times)) * 1000,
        "min_ms": min(times) * 1000,
        "max_ms": max(times) * 1000,
        "std_ms": (np.std(times) * 1000) if len(times) > 1 else 0,
        "result": float(result),  # For cross-validation
    }


def main():
    print(f"NumPy sum() Benchmark")
    print(f"NumPy version: {np.__version__}")
    print(f"Array sizes: {', '.join(f'10^{len(str(s))-1}' for s in ARRAY_SIZES)}")
    print("-" * 50)

    results = {
        "library": "numpy",
        "version": np.__version__,
        "config": {
            "warmup_iterations": WARMUP_ITERATIONS,
            "min_iterations": MIN_ITERATIONS,
            "min_duration_sec": MIN_DURATION_SEC,
        },
        "benchmarks": []
    }

    for size in ARRAY_SIZES:
        print(f"Benchmarking size {size:>10,}...", end=" ", flush=True)
        result = benchmark_sum(size)
        results["benchmarks"].append(result)
        print(f"Mean: {result['mean_ms']:>10.4f} ms  ({result['iterations']} iterations)")

    # Output to JSON
    output_dir = Path(__file__).parent.parent / "results"
    output_dir.mkdir(exist_ok=True)
    output_file = output_dir / "numpy_results.json"

    with open(output_file, "w") as f:
        json.dump(results, f, indent=2)

    print("-" * 50)
    print(f"Results saved to: {output_file}")


if __name__ == "__main__":
    main()
